s3access/s3async/s3select.py

```diff
@@ -0,0 +1,144 @@
+        raise IOError("stream was not read in full")
+    return res
+
+
+def _select_tasks(
+        client: AioBaseClient,
+        sources: Sources,
+        expression: str,
+        expression_type: ExpressionType = ExpressionType.SQL,
+        input_serialization: Input = DEFAULT_INPUT,
+        output_serialization: Output = DEFAULT_OUTPUT,
+        scan_range: ScanRange = DEFAULT_SCAN_RANGE,
+        progress_callback: Optional[ProgressCallback] = None) -> Tuple[Sequence[Coroutine], Sequence[BucketKeyPair]]:
+    tasks = []
+    key_list = []
+    for bucket, keys in sources.items():
+        for key in keys:
+            tasks.append(
+                select(client, bucket, key, expression, expression_type, input_serialization,
+                       output_serialization, scan_range, progress_callback))
+            key_list.append((bucket, key))
+    return tasks, key_list
+
+
+async def multiple(
+        client: AioBaseClient,
+        sources: Sources,
+        expression: str,
+        expression_type: ExpressionType = ExpressionType.SQL,
+        input_serialization: Input = DEFAULT_INPUT,
+        output_serialization: Output = DEFAULT_OUTPUT,
+        scan_range: ScanRange = DEFAULT_SCAN_RANGE,
+        progress_callback: Optional[ProgressCallback] = None) -> Dict[BucketKeyPair, bytes]:
+    tasks, key_list = _select_tasks(
+        client, sources, expression, expression_type, input_serialization,
+        output_serialization, scan_range, progress_callback)
+    results = await asyncio.gather(*tasks)
+    return dict(zip(key_list, results))
+
+
+async def multiple_as_completed(
+        client: AioBaseClient,
+        sources: Sources,
+        expression: str,
+        expression_type: ExpressionType = ExpressionType.SQL,
+        input_serialization: Input = DEFAULT_INPUT,
+        output_serialization: Output = DEFAULT_OUTPUT,
+        scan_range: ScanRange = DEFAULT_SCAN_RANGE,
+        progress_callback: Optional[ProgressCallback] = None) -> AsyncIterator[Tuple[bytes, BucketKeyPair]]:
+    tasks, key_list = _select_tasks(
+        client, sources, expression, expression_type, input_serialization,
+        output_serialization, scan_range, progress_callback)
+    for i, file_f in enumerate(asyncio.as_completed(tasks)):
+        file = await file_f
+        yield file, key_list[i]
+
+
+async def parquets_to_csv_bytes(
+        client: AioBaseClient,
+        sources: Sources,
+        columns: Optional[Sequence[str]] = None,
+        filters: Sequence[Tuple[Any, str, Any]] = None,
+        progress_callback: Optional[ProgressCallback] = None) -> bytes:
+    """
+    example usage:
+    -----------------------------------------
+    def print_progress(bucket, key, details):
+        if details == 'completed':
+            print(bucket, key, details)
+    bucket = 'some-bucket'
+    keys = ['key/key1.parquet', 'key/key2.parquet', ...]
+    columns = ['col1', 'col2', 'col5', 'col10']  # some columns all files share
```

Comment:
How are the types resolved?

Reply:
s3select (async) does not provide type resolution as it is providing the lowest level it can for selecting data out of s3  (with a bit of specialization in parquet selection).
In my opinion this allows for better compos-ability with other/different code (hence the pandas specialization file).
one can choose simply to dump the csv blob, read it using csv package or whatever else he wishes to do with bytes.

Reply:
fair enough.